### ISSUE TITLE
Snowflake Destination: use better file size with S3 staging files

### ISSUE THREAD

**Issue by marcosmarxm** [airbytehq/airbyte#6517]
## Tell us about the problem you're trying to solve
From [slack convo](https://airbytehq.slack.com/archives/C019WJFR1JT/p1632496795259100)

Use a better 

https://docs.snowflake.com/en/user-guide/data-load-considerations-prepare.html#general-file-sizing-recommendations

> 100-250MB(compressed) per file is recommended but can go less depends on the total data size. If total size was 100MB, you could split it in to smaller files 10 x 10MB or 5 x 20MB. Ingestion process uses one core per file and each node has 8 cores and each cluster can have 1-128 nodes depending on the size. This means each node can ingest 8 files at a time. I would use logic where total amount of data is split in to as many smaller files as possible where each file size 10 to 250MB each. This will ensure all the cores in the cluster will have something to process w/o sitting idle. (edited) 

> AS for where the files are exported for ingestion there should be 2 options with following sub-selecttions.
External Stage (meaning customer has blob storage of their own that they manage) where your code would do the job of uploading files to appropriate cloud storage locations.
AWS (credentials + S3 path )
Azure (credentials + blob path )
GCP Storage (credentials + blob path )
Internal Stage (Snowflake managed storage defined by a name where access is granted via Snowflake RBAC to userid that is making connection to Snowflake). You would use JDBC, ODBC or another Snowflake driver to connect snowflake and issue a PUT SQL command that would encrypt & upload the file(s) automatically. You do not have to code the upload process in your connector and it is handled by the ODBC/JDBC driver.
Option 1, use an existing Internal Stage: In this case, you would only define the Schema + the name of the internal stage along with optional  SubFoldering to use.
Option 2, use temp internal stage. user would specify a schema to use. You can use SQL CREATE STAGE to create a upload location, SQL PUT to upload local files to the stage & COPY to ingest them. Once data is ingested, You can delete the stage and that will get rid of all the files in that stage. (most ELT tools use this option)
https://docs.snowflake.com/en/user-guide/data-load-local-file-system-stage.html
https://docs.snowflake.com/en/user-guide/data-load-local-file-system-create-stage.html#creating-a-stage

**Comment by andriikorotkov** [comment]
Hello, @misteryeo, @marcosmarxm, @alexandr-shegeda. I was able to split the files for Snowflake s3 staging into smaller ones (about 120-140 MB in size). I added a pull request link with these changes to this task. It is currently in draft form.

Tell me, please, is it necessary to add such changes to work with snowflake GCS staging files? If so, I would like to reuse an already existing destination - destination-GCS. To do this, I will need to change the spec that is currently used for snowflake GCS staging to the one that we use for destination-GCS. In this case, users who used the old version snowflake with GCS staging will need to update the specification . What do you say about this?

**Comment by sherifnada** [comment]
@andriikorotkov does splitting the size help performance?

**Comment by andriikorotkov** [comment]
@sherifnada, based on the documentation, reducing the size of the staging file to the optimal size of 100-250 MB should improve performance. Also, I've added a writer closure after the desired amount has been written to the staging file. With this change, we will be closing writers sequentially, rather than keeping them open, throughout the entire time all data is being written to the staging files. This insures us against OOM exceptions.
In terms of performance, I tested migrating data from MySQL database (6 tables, 5M records each, a total of just over 12 GB of data) to the snowflake. I did not notice significant improvements in the speed of work. The speed of work before changes is approximately 2 hours 30 minutes. The speed of work after the changes made is approximately 2 hours 10 minutes.
